src/Catalog/Icons/IconsCollector.cs

```diff
@@ -0,0 +1,121 @@
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using NuGet.Protocol.Catalog;
+using NuGet.Services.Metadata.Catalog.Persistence;
+
+namespace NuGet.Services.Metadata.Catalog.Icons
+{
+    public class IconsCollector : CommitCollector
+    {
+        private readonly IStorageFactory _targetStorageFactory;
+        private readonly ICatalogClient _catalogClient;
+        private readonly ICatalogLeafDataProcessor _catalogLeafDataProcessor;
+        private readonly IIconCopyResultCachePersistence _iconCopyResultCache;
+        private readonly ILogger<IconsCollector> _logger;
+
+        public IconsCollector(
+            Uri index,
+            ITelemetryService telemetryService,
+            IStorageFactory targetStorageFactory,
+            ICatalogClient catalogClient,
+            ICatalogLeafDataProcessor catalogLeafDataProcessor,
+            IIconCopyResultCachePersistence iconCopyResultCache,
+            Func<HttpMessageHandler> httpHandlerFactory,
+            ILogger<IconsCollector> logger)
+            : base(index, telemetryService, httpHandlerFactory, httpClientTimeout: TimeSpan.FromMinutes(5))
+        {
+            _targetStorageFactory = targetStorageFactory ?? throw new ArgumentNullException(nameof(targetStorageFactory));
+            _catalogClient = catalogClient ?? throw new ArgumentNullException(nameof(catalogClient));
+            _catalogLeafDataProcessor = catalogLeafDataProcessor ?? throw new ArgumentNullException(nameof(catalogLeafDataProcessor));
+            _iconCopyResultCache = iconCopyResultCache ?? throw new ArgumentNullException(nameof(iconCopyResultCache));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        protected override Task<IEnumerable<CatalogCommitItemBatch>> CreateBatchesAsync(
+            IEnumerable<CatalogCommitItem> catalogItems)
+        {
+            var maxCommitTimestamp = catalogItems.Max(x => x.CommitTimeStamp);
+
+            return Task.FromResult<IEnumerable<CatalogCommitItemBatch>>(new[]
+            {
+                new CatalogCommitItemBatch(
+                    catalogItems,
+                    key: null,
+                    commitTimestamp: maxCommitTimestamp),
+            });
+        }
+
+        protected override async Task<bool> OnProcessBatchAsync(
+            CollectorHttpClient client,
+            IEnumerable<CatalogCommitItem> items,
+            JToken context,
+            DateTime commitTimeStamp,
+            bool isLastBatch,
+            CancellationToken cancellationToken)
+        {
+            await _iconCopyResultCache.InitializeAsync(cancellationToken);
+
+            var filteredItems = items
+                .GroupBy(i => i.PackageIdentity)                          // if we have multiple commits for the same package (id AND version)
+                .Select(g => g.OrderBy(i => i.CommitTimeStamp).ToList()); // group them together for processing in order
```

Comment:
Can you leave a comment about why you process all leafs per package? Why not just the latest? I think we talked offline about this but it is worth commenting on.